README
README

December 2009
Author:
    Terence Stenvold <tstenvold@gmail.com>
Contributors:
    Carson Farrell

Special Notes
=============
The goal of this project is really to gain as many MAC address mapped to device vendors as possible
If you see an entry on the CLI stating that the Vendor is 'Unknown". Harald Scan probably created a file in the
same directory with the first 6 characters from the MAC address with all the information it could gain from the
device. This will allow me to add it to the MACLIST if I am able to determine who the Vendor is. Eventually this
may be an automatic function but right now I request you email me these files @  tstenvold@gmail.com


Installation
============
1. Unpack to a directory
2. May need to chmod +x haraldscan (as root)
3. Run './haraldscan -b' to build database
4. You are ready to go

Usage
======

Usage: haraldscan [options]

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -b, --build           Builds MAC Addr database. Ignores all other options
  -f FLUSHNUM, --flush=FLUSHNUM
                        When db = size entered. Flush entries to a different
                        database (useful if combined with -m)
  -m, --memorydb        Puts the database in Memory instead of a file on disk
  --no-service          Disables service scans on 'Unknown' devices
  --no-write            Disables writing discovered device info to a file
  -s, --service         Does a service scan of all devices found and saves a
                        file like a 'Unknown' device would
  -t NMINUTES, --time=NMINUTES
                        Shows number of devices found per time specified in
                        mins (default is 15 mins)
  -u, --update          Updates the MACLIST if there are updates and rebuilds
                        the database (requires and Internet connection)
  -w FILENAME, --write=FILENAME
                        Outputs discovered device info to a file you specify
                        (unspecified: filename is a timestamp)


Changelog.md
# 0.42 #

## Distribution ##
  * 32 bit and 64 bit binary for Linux
  * OS X version still source code

## Output ##
  * In XML format for easier parsing

## MACLIST ##
  * Updated to 345 Entries.

# 0.41 #

## Distribution ##
  * 32 bit binary will be used for distribution
  * Source code will be labeled as such

## Interface ##
  * Added a coloured title
  * Fixed some displaying issues

## MACLIST ##
  * Updated to 310 Entries.

# 0.401 #
  * There is no diffence in functionality between 0.4 and 0.401

## Distribution ##
  * Revered back to just a src dist. The binary seems to not be working.

## Changed Features ##
  * In the archive the files are now in a folder again.


# 0.4 #

## Distribution ##
  * ~~Harald Scan is now distributed in either source code or dist~~
  * Harald Scan is still distributed by means of source code
## Added Features ##
  * CLI has added a Number of devices discovered by duration (default is 15 mins)
  * The duration for above is configurable through `-t TIME`
  * You can choose either a in memory database or a file database
  * Option to disable services scans entirely even if MAC is 'Unknown'
  * Option to disable writing devices to a text file
  * ability to check Harald Scan version with `--version`
  * Added `-f` for flushing a large database to another db file. Typical usage would be with a large in memory database.

## Changed Features ##
  * Harald Scan will write to a file every time unless `--no-write` is specified
  * Changed argument handling to optparse much better now
  * Uses sqlite3 module instead of pysqlite2 (I didn't realize it was there :S)

## Bug Fixes ##
  * A bunch don't remember though


haraldargs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# haraldargs.py
# July 2009
# Terence Stenvold <tstenvold@gmail.com>
#
#Command Line Args handling

#This file is part of Haraldscan.
#
#Haraldscan is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License Version 3 as
#published by the Free Software Foundation.
#
#Haraldscan is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License Version 3 for more details.
#
#You should have received a copy of the GNU General Public License
#Version 3 along with Haraldscan.  If not, see <http://www.gnu.org/licenses/>.

import haraldusage
import haraldsql
import haraldupdate
import sys, time
from optparse import OptionParser

def cmd_parse(argv):

    parser = OptionParser(usage="usage: haraldscan.py [options]", version="%prog 0.4")

    parser.add_option("-b", "--build",
                      action="store_true",
                      dest="build",
                      default=False,
                      help="Builds MAC Addr database. Ignores all other options")
    parser.add_option("-f", "--flush",
                      action="store",
                      type="int",
                      dest="flushnum",
                      default=0,
                      help="When db = size entered. Flush entries to a different database (useful if combined with -m)")
    parser.add_option("-m","--memorydb",
                      action="store_true",
                      dest="memdb",
                      default=False,
                      help="Puts the database in Memory instead of a file on disk")
    parser.add_option("--no-service",
                      action="store_true",
                      dest="noservice",
                      default=False,
                      help="Disables service scans on \'Unknown\' devices")
    parser.add_option("--no-write",
                      action="store_true",
                      dest="nowrite",
                      default=False,
                      help="Disables writing discovered device info to a file")
    parser.add_option("-s", "--service",
                      action="store_true",
                      dest="service",
                      default=False,
                      help="Does a service scan of all devices found and saves a file like a 'Unknown' device would")
    parser.add_option("-t", "--time",
                      action="store",
                      dest="nminutes",
                      type="int",
                      default=15,
                      help="Shows number of devices found per time specified in mins (default is 15 mins)")
    parser.add_option("-u", "--update",
                      action="store_true",
                      dest="update",
                      default=False,
                      help="Updates the MACLIST if there are updates and rebuilds the database (requires and Internet connection)")
    parser.add_option("-w", "--write",
                      action="store",
                      type="string",
                      dest="filename",
                      default=str(time.time()),
                      help="Outputs discovered device info to a file you specify (unspecified: filename is a timestamp)")

    return parser

def handle_args(argv,c):

    parser = cmd_parse(argv)

    (options, args) = parser.parse_args()

    c.service = options.service
    c.time_interval = options.nminutes
    c.buildb = options.build
    c.noservice = options.noservice
    c.memdb = options.memdb
    c.flush = options.flushnum

    if options.nowrite is False:
        c.minus_w(options.filename)

    if options.update is True:
        if haraldupdate.check_now():
            c.buildb = True
        else:
            sys.exit(0)

    if haraldsql.chk_database() == False and c.buildb == False:
        haraldusage.no_db()

if __name__ == "__main__":
    parser = cmd_parse([""])
    parser.print_help()
